(Коммит для пул реквеста) Задачи по теме: "Классы, интерфейсы, пакеты…

ml-rdv opened the pull request:
…, модули"

ml-rdv committed:
(Коммит для пул реквеста) Задачи по теме: "Классы, интерфейсы, пакеты, модули"

ml-rdv committed:
Удаление для дальнейшего добавления и пул реквест CorrectionsForTasks

ml-rdv committed:
Добавление задач для пул реквеста в ветке CorrectionsForTasks

Adgam commented on src/PracticeClassesInterfacesPackagesModules/MyLinkedListNodeIsNestedClass/MyLinkedListNodeIsNestedClass.java:
```diff
@@ -0,0 +1,156 @@
+ */
+public class MyLinkedListNodeIsNestedClass {
+
+    public static class Node {
```
Как думаешь для данного случая должен ли класс Node быть публичным?

Adgam commented on src/PracticeClassesInterfacesPackagesModules/MyLinkedListIsGeneric/Node.java:
```diff
@@ -0,0 +1,37 @@
+package PracticeClassesInterfacesPackagesModules.MyLinkedListIsGeneric;
+
+public class Node {
+    private final Object data;
```
Хорошо бы, чтобы это тоже был generic type

Adgam commented on src/PracticeClassesInterfacesPackagesModules/DeepCopyWithCloneable/DeepCopyWithCloneableTest.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @Test
+    void orders_should_be_equals() {
```
не совсем понял зачем это тест, он же не тестирует логику клонирования

Adgam commented on src/PracticeClassesInterfacesPackagesModules/DeepCopyWithSerialization/DeepCopyWithSerializationTest.java:
```diff
@@ -0,0 +1,49 @@
+        Assertions.assertNotEquals(order.client.phoneNumber, orderClone.client.phoneNumber);
+    }
+
+    public static Object deepCopy(Object object) {
```
забежала маленько вперёд) 
ну вариант хороший, сможешь дать сравнение клонирования через cloneable и сериализацию?

Adgam commented on src/PracticeClassesInterfacesPackagesModules/DeepCopyWithSerialization/DeepCopyWithSerializationTest.java:
```diff
@@ -0,0 +1,49 @@
+
+    public static Object deepCopy(Object object) {
+        try {
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```
Стримы нужно закрывать через вызов метода close(), либо через try-catch-resources, иначе есть риск утечки памяти, но это отдельная тема, скоро до неё дойдём.

Adgam commented on src/PracticeClassesInterfacesPackagesModules/WorkWithEnumAndRecord/MethodGetsJustOneEnum.java:
```diff
@@ -0,0 +1,91 @@
+        Season s = null;
+        int index = 0;
+        String name = null;
+        switch (month) {
```
попробуй сделай вычисление программно, тут повезло, что месяцев 12 и это вряд ли когда изменится) но вот с другими данными это уже может не прокатить. 

Время года, номер месяца и даже название на русском в именительном падеже можно вычислить не прибегая к хардкоду. С количеством дней в году у тебя хорошо получилось, можешь двигаться примерно в том же направление.

Adgam commented on src/PracticeClassesInterfacesPackagesModules/WorkWithEnumAndRecord/MethodGetsJustOneEnum.java:
```diff
@@ -0,0 +1,91 @@
+                index = 12;
+                name = "Декабрь";
+            }
+            default -> System.out.println("It is impossible date.");
```
данный кейс возможен? если ты попробуешь передать в метод что-то помимо месяца из enum, что произойдёт?
Или это защита от дурака, на случай если случайно добавят 13 месяц? тогда я бы здесь поставил выброс исключения

Adgam commented on src/PracticeClassesInterfacesPackagesModules/WorkWithEnumAndRecord/WorkWithEnumAndRecordTest.java:
```diff
@@ -0,0 +1,70 @@
+    void information_about_every_month_should_be_right(Month month) {
+        MonthWithInfo monthWithInformation = MethodGetsJustOneEnum.makeRecord(month);
+        MonthWithInfo expectedElement = null;
+        switch (month) {
```
от switch лучше избавиться, тест слишком перегруженный получается, слишком много потенциальных разветвлений, тут либо разбить на 12 тестов или сделать что-то вроде



```java
    private static List<Arguments> provideParameters() {
        return List.of(
                Arguments.of(January, new MonthWithInfo(Season.Winter, 31, 1, "Январь")),
                Arguments.of(February, new MonthWithInfo(Season.Winter, 28, 2, "Февраль")),
                Arguments.of(March, new MonthWithInfo(Season.Spring, 31, 3, "Март")),
                Arguments.of(April, new MonthWithInfo(Season.Spring, 30, 4, "Апрель")),
                Arguments.of(May, new MonthWithInfo(Season.Spring, 31, 5, "Май")),
                Arguments.of(June, new MonthWithInfo(Season.Summer, 30, 6, "Июнь")),
                Arguments.of(July, new MonthWithInfo(Season.Summer, 31, 7, "Июль")),
                Arguments.of(August, new MonthWithInfo(Season.Summer, 31, 8, "Август")),
                Arguments.of(September, new MonthWithInfo(Season.Autumn, 30, 9, "Сентябрь")),
                Arguments.of(October, new MonthWithInfo(Season.Autumn, 31, 10, "Октябрь")),
                Arguments.of(November, new MonthWithInfo(Season.Autumn, 30, 11, "Ноябрь")),
                Arguments.of(December, new MonthWithInfo(Season.Winter, 31, 12, "Декабрь"))
        );
    }

    @ParameterizedTest
    @MethodSource("provideParameters")
    void information_about_every_month_should_be_right(Month month, MonthWithInfo monthWithInfo) {
        MonthWithInfo monthWithInformation = MethodGetsJustOneEnum.makeRecord(month);
        Assertions.assertEquals(monthWithInfo, monthWithInformation);
    }
```

ты можешь спросить, а какая по сути разница, почти тоже самое, я же отвечу, что исключён момент определения ожидаемого результата, т.е. по факту запуститься также 12 тестов, но в тест будет чётко подаваться месяц и ожидаемая информация о нём

Adgam commented on src/PracticeClassesInterfacesPackagesModules/WorkWithEnumAndRecord/WorkWithEnumAndRecordTest.java:
```diff
@@ -0,0 +1,70 @@
+                expectedElement = new MonthWithInfo(Season.Winter, 31, 1, "Январь");
+            }
+            case February -> {
+                int daysInMonth = Year.now().getValue() % 4 == 0 ? 29 : 28;
```
потенциальное опасное место, проблема та же, что и в предыдущих задачах с датами, хорошо бы вообще отдельный тест на это написать и реализовать один из приёмов, которые мы разбирали для решения данной проблемы, но в данной задаче предлагаю забить, просто обращаю твоё внимание

ml-rdv committed:
Changes in solutions
1. Nested class Node must be private
2. Class Node must be Generic
3. Stream must be closed
4. Class MethodGetsJustOneEnum must be optimized

ml-rdv committed:
Exceptions
DividingTwoIntegers
DividingTwoDoubles

Adgam commented on src/PracticeClassesInterfacesPackagesModules/WorkWithEnumAndRecord/MethodGetsArrayOfEnums.java:
```diff
@@ -31,74 +33,18 @@ public static <T extends Enum<T>> ArrayList<MonthWithInfo> checkMonth(T[] values
-            }
-            default -> System.out.println("It is impossible date.");
-        }
+        int index = Month.valueOf(month.name()).ordinal() + 1;
```
почему не просто?
`int index = month.ordinal() + 1;`

Adgam commented on src/PracticeClassesInterfacesPackagesModules/WorkWithEnumAndRecord/MethodGetsArrayOfEnums.java:
```diff
@@ -31,74 +33,18 @@ public static <T extends Enum<T>> ArrayList<MonthWithInfo> checkMonth(T[] values
+
+        java.time.Month monthCopy = java.time.Month.of(index);
+        Locale loc = Locale.forLanguageTag("ru");
+        String name = monthCopy.getDisplayName(TextStyle.FULL_STANDALONE, loc);
```
ещё есть вариант получить названия месяца
`yearMonthObject.format(DateTimeFormatter.ofPattern("MMMM"));`
но это будет родительный падеж, если надо именительный, то
`yearMonthObject.format(DateTimeFormatter.ofPattern("LLLL"));`

Adgam commented on src/PracticeClassesInterfacesPackagesModules/WorkWithEnumAndRecord/MethodGetsArrayOfEnums.java:
```diff
@@ -31,74 +33,18 @@ public static <T extends Enum<T>> ArrayList<MonthWithInfo> checkMonth(T[] values
+        String name = monthCopy.getDisplayName(TextStyle.FULL_STANDALONE, loc);
+        name = name.substring(0, 1).toUpperCase() + name.substring(1);
+
+        Season s = Season.of(month);
```
пока не буду сильно цепляться к названию переменных, но на пет проджектах буду душнить))

Adgam commented on src/PracticeClassesInterfacesPackagesModules/WorkWithEnumAndRecord/MethodGetsJustOneEnum.java:
```diff
@@ -17,75 +19,27 @@
+        String name = monthCopy.getDisplayName(TextStyle.FULL_STANDALONE, loc);
+        name = name.substring(0, 1).toUpperCase() + name.substring(1);
+
+        Season s = seasons[index - 1];
```
массив с времена года выглядит как костыль, решение через switch уже лучше)

я тебе намекал в прошлый раз на подобное решение, разберись плиз в нём.
```java
int monthNumInSeason = 3;
int seasonNum = Season.values().length;
int seasonIndex = (index / monthNumInSeason) % seasonNum;

Season s = Season.values()[seasonIndex];
```

Тут я не делаю акцент, что это более правильное решение чем твои, просто показываю как ещё можно было сделать, поставленную задачу ты выполнила, времена года вычисляются, поэтому зачёт)

Adgam commented on src/PracticeClassesInterfacesPackagesModules/WorkWithEnumAndRecord/Season.java:
```diff
@@ -1,8 +1,21 @@
+    static public Season of(final Month month) {
+        Objects.requireNonNull(month, "ERROR - Received null where a `Month` is expected.");
+        return
+                switch (Objects.requireNonNull(month)) {
```
здесь `Objects.requireNonNull` избыточный, т.к. `Objects.requireNonNull`, который выше выбросит исключение в случае null и мы не дойдём до свича

Adgam commented on src/Exceptions/dividingTwoDoubles/dividingTwoDoubles.java:
```diff
@@ -0,0 +1,23 @@
+                throw new ArithmeticException("Double overflowed");
+            }
+        } catch (NullPointerException ex) {
+            throw new IllegalArgumentException("Number is null");
```
проверку на null и выбрасывание исключения IllegalArgumentException лучше делать в начале метода, 
во-первых, ты здесь перехватываешь абсолютно все исключения NullPointerException, а это не всегда может значить, что у тебя аргументы == null.
во-вторых, аргументы лучше провалидровать сначала и выбросить исключение, не допуская выполнения логики, т.к. возможно неявное поведение

Adgam commented on src/Exceptions/dividingTwoDoubles/dividingTwoDoubles.java:
```diff
@@ -0,0 +1,23 @@
+        try {
+            result = firstNum / secondNum;
+            System.out.println(result);
+            if (result == Double.POSITIVE_INFINITY) {
```
POSITIVE_INFINITY может быть не только при деление на ноль, например при `Double.MAX_VALUE + Double.MAX_VALUE`, в этом случае у тебя будет ошибка. 

Проверка на ноль лучше делать сразу в начале метода и не допускать до выполнения основной логики.

\+ Нужно выбрасывать **собственное** checked исключение, т.е. нужно создать свой класс-исключение.

ml-rdv committed:
Ппроверку на null и выбрасывание исключения IllegalArgumentException в начале метода
Собственное исключение DividerIsNull, вызывается, когда делитель = 0

ml-rdv committed:
Задачи по теме "Коллекции"
1. Для ArrayList написать тесты на методы (обратить внимание на перегрузки и различные вариации по типу add/addAll или remove/removeIf)
Добавления элементов
Замены элементов
Получение элементов
Удаление элементов
Определение признака, что элемент содержится в списке
Получение размера и признака, что список пуст
2. Для HashMap написать тесты на методы по аналогии с предыдущим заданием
3. Запустить код и объяснить почему hashmap не отдаёт значение при втором вызове метода get()
4. Как в данном примере сделать так, что Person в hashSet стал уникальным?
5. Выбрать класс из задачи по ООП и реализовать в нём интерфейс Comparable. Затем написать тест в котором сортируется список с объектами класса из задачи по ООП

ml-rdv committed:
Задачи по теме "Коллекции"
1. Для ArrayList написать тесты на методы (обратить внимание на перегрузки и различные вариации по типу add/addAll или remove/removeIf)
Добавления элементов
Замены элементов
Получение элементов
Удаление элементов
Определение признака, что элемент содержится в списке
Получение размера и признака, что список пуст
2. Для HashMap написать тесты на методы по аналогии с предыдущим заданием
3. Запустить код и объяснить почему hashmap не отдаёт значение при втором вызове метода get()
4. Как в данном примере сделать так, что Person в hashSet стал уникальным?
5. Выбрать класс из задачи по ООП и реализовать в нём интерфейс Comparable. Затем написать тест в котором сортируется список с объектами класса из задачи по ООП

Adgam commented on src/Exceptions/dividingTwoDoubles/DividerIsNull.java:
```diff
@@ -0,0 +1,7 @@
+package Exceptions.dividingTwoDoubles;
+
+public class DividerIsNull extends Exception {
```
Обычно исключения принято называть по типу DividerIsNullException

ml-rdv replied:
@Adgam 
fixed

Adgam commented on src/Collections/MutableKeyInHashMap.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Запустить код и объяснить почему hashmap не отдаёт значение при втором вызове метода get()
+ * <p>
+ * Переменная num влияет на формирование хэшкода.
```
Не хватило деталей в объяснение, обсудим на занятие

ml-rdv committed:
Использовать TreeSet, который содержит числа отсортированные по убыванию

Adgam commented on src/Collections/ClientComparableRealization.java:
```diff
@@ -0,0 +1,48 @@
+ * Выбрать класс из задачи по ООП и реализовать в нём интерфейс Comparable.
+ * Затем написать тест в котором сортируется список с объектами класса из задачи по ООП
+ */
+public class ClientComparableRealization implements Comparable<ClientComparableRealization> {
```
Обсудить понимание Comparable

Adgam commented on src/Collections/ComparatorRealization.java:
```diff
@@ -0,0 +1,24 @@
+        String fullName1 = a == null ? null : a.getFullName();
+        String fullName2 = b == null ? null : b.getFullName();
+
+        return fullName1.compareTo(fullName2);
```
Я так понимаю здесь NullPointerException вылетает. Если пришёл null, то можно пустую строку подставлять для fullName как вариант или определить порядок, где все null должны находиться вверху или внизу, но для этого надо условия добавить

ml-rdv replied:
@Adgam 
fixed

Adgam commented on src/Collections/TreeSetWithReverseSort.java:
```diff
@@ -0,0 +1,21 @@
+        // Вариант 2 без создания отдельного класса, с использованием лямбда-выражения
+        // Comparator<Integer> comp = (Integer o1, Integer o2) -> (o2 - o1);
+
+        final TreeSetComparator comparator = new TreeSetComparator();
```
можно просто `new TreeSet<>(Comparator.reverseOrder())` без создания отдельного компоратора

ml-rdv committed:
Корректировки по комментариям

ml-rdv committed:
Корректировки по комментариям (Обычно исключения принято называть по типу DividerIsNullException)

ml-rdv committed:
Создать свой класс Comparator для класса из задачи по ООП и также написать тест на сортировку
В список нужно добавить null’ы и сделать так, чтобы сортировка могла их  обработать

ml-rdv committed:
Корректировки

ml-rdv committed:
Корректировки
